### PR TITLE
New post:" Unlock the power of Unstructured Data: From Embeddings to In Context Learning – Build a Full stack Q&A Chatbot with Langchain, and LLM Models on Sagemaker

### PR DESCRIPTION
**Pull request by pkamra**
# Content Pull Request Checklist
Initial Draft of Content for Build On AWS Tutorial  - "Unlock the power of Unstructured Data: From Embeddings to In Context Learning – Build a Full stack Q&A Chatbot with Langchain, and LLM Models on Sagemaker"
The Approved Asana Ticket for this tutorial is https://app.asana.com/0/1204199791673171/1205326498728997

Please check to make sure your content submission meets the following requirements by adding an `x` between the brackets:

- [ Y] The title of this pull request reflects the title of my content.
- [ Y] I have reviewed the [Content Review Checklist](/content/blob/main/CONTENT_REVIEW_CHECKLIST.md).
- [Y ] This pull request contains one piece of content (post or page).

**Important: if you are submitting a PR to update existing content, please ensure you [sync'ed your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) first before you created your branch.**

Is this pull request linked to a Content Proposal or Idea Suggestion Issue? If so, please link to it below using the number or full link.

```text
Resolves #5

or

Content for https://github.com/build-on-aws/content/issues/5
```

**By submitting this pull request, I confirm I own this content, have permission to use each image included in this content, and that it will be released under the [CC BY-SA 4.0 SA License](/LICENSE).**

### CODE REVIEW

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+curl -X POST -H "Content-Type: application/json" -d '{"question":"How can I clean my windshield?"}' http://quest-Publi-abc-xxxx.us-east-1.elb.amazonaws.com/qa
+```
+You will see a response as shown below 
+```bash 
+{"response":"To clean sensors and camera lenses, use a cloth moistened with a small amount of glass detergent."}
+```
+
+## Step 7 - Build and deploy a Website with an embedded Chatbot that is integrated to our API.
+
+Next we move on to the last step for our full stack pipeline, which is integrating the API with our Open Source embedded chatbot using Open Source botkit in a HTML website. For this website and the embedded chatbot, our source code is a node based application consisting of an index.html integrated with Open Source botkit.js as the chatbot. To make things easy I have created a Dockerfile and provided it alongside the code in the folder `homegrown_website_and_bot`. Once again, build the Docker container and push the image to AWS ECR repository.
```
Comment (abhirockzz) [inline tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md:146]: fix "node based application" - Nodejs application

**Review by abhirockzz — CHANGES_REQUESTED** [review]
This is good topic and shows an end-to-end solution. However, I have the following concerns:

- At this point, it's not fit to be published as a tutorial - it does not provide step-by-step instructions. A tutorial is something that reader can complete.
- Lot of missing context. Concepts and services have not been introduced.
- Way too many diagrams. It's partly because they are not being used properly e.g. don't explain code using screenshots - refer to the GitHub repo and explain it in the content itself

This piece needs to be significantly overhauled to qualify as a tutorial. Also needs improvement in terms of readability and structure

**Comment by pkamra** [comment]
Need clarification on the below comment since it doesnt tell me what to fix ?
>> At this point, it's not fit to be published as a tutorial - it does not provide step-by-step instructions. A tutorial is something that reader can complete.

I have provided a Step by Step instruction to build ? Which part was not reproducible or which part do I need to change ?

**Comment by pkamra** [comment]
Need clarification on this. Let me know which service needs explanation or if you could give me an example taht woudl help.
>> Lot of missing context. Concepts and services have not been introduced.

**Comment by pkamra** [comment]
I will remove the screen shots showing the code explanation to the code repo.
>> Way too many diagrams. It's partly because they are not being used properly e.g. don't explain code using screenshots - refer to the GitHub repo and explain it in the content itself

**Comment by pkamra** [comment]
working on the feedback.

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
+---
+title: "Unlock the power of Unstructured Data: From Embeddings to In Context Learning – Build a Full stack Q&A Chatbot with Langchain, and LLM Models on Sagemaker"
```
[the remark this answers is not in the document]
Comment (pkamra) [reply tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md]: Updated

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+![Save embeddings in OpenSearch 1](images/opensearch-embeddings-1.jpg)
+![Save embeddings in OpenSearch 3](images/opensearch-embeddings-3.jpg)
+
+## Step 6 - Build and deploy an API for providing real time "context" to the LLM and retrieve back meaningful responses.
+
+Now that we have our text embeddings in the Vector Database powered by AWS OpenSearch, let us move to the next step where we will utilize the T5 Flan XXL Large Language Model (LLM) to respond to inquiries about our car manual. We will use our saved embeddings in the vector database to <b>provide context</b> to the LLM and use the LLM's capabilities to interpret text to answer questions that we ask the LLM about our car manual. To implement our use case, we will rely heavily on Open Source LangChain. LangChain is a Python framework that makes it easy to orchestrate and combine the various components that we need to create our real-time <b>context aware question-answer system powered by our LLM</b>. LangChain makes it easy for us to inject our car manual content into our prompt template, and send the final prompt to our LLM model.
```
[the remark this answers is not in the document]
Comment (pkamra) [reply tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md]: Updated Step 5  and Step 6 with detailed steps.

**Comment by pkamra** [comment]
@abhirockzz  have updated the content. Removed screen shots, added detailed steps where missing, Please review.

**Comment by abhirockzz** [comment]
Thanks for incorporating the feedback. I do have concerns about the diagrams. Please fix them and we should be good to go:

General comment on diagrams - Most of the them need to be cleaned up 

- Use proper/complete service names and updated logos (refer to https://aws.amazon.com/architecture/icons/),
- arrows are overlapping with text
- inconsistent services logos have been used (e.g. in some cases I see Fargate with a black background, and in some cases its different). 
- Lambda pipeline diagram - pls remove text from the diagram. Use appropriate logo/icons
- pls don't include too much text in the diagram (it's distracting). Leave that for the article/content
- not sure why a green background has been chosen for diagrams. It's not suitable both from readability and aesthetic point of view
- some of the diagrams are not clear. For e.g. in the SageMaker studio diagram, I can hardly see anything (I am familiar with SageMaker studio, but a new reader might have a tough time). Not sure if this is a platform issue or something to do with the image(s)

Also:

- Clean up section - remove the manual steps and provide CLI instructions.
- Add a "Prerequisites" section with all the tooling required to complete the tutorial (e.g. AWS CLI, Docker etc.)

**Comment by cherbk** [comment]
> Working on fixing the diagrams

I will submit it for editorial review while you work on the diagrams. Thanks Piyali

**Comment by pkamra** [comment]
@abhirockzz and @cherbk, Have incorporated the feedback for diagrams, added additional details in Prerequisites under the Attributes table on top and  CLI Instructions in the Cleanup section.

**Review by abhirockzz — APPROVED** [review]

**Comment by abhirockzz** [comment]
> @abhirockzz and @cherbk, Have incorporated the feedback for diagrams, added additional details in Prerequisites under the Attributes table on top and CLI Instructions in the Cleanup section.

Thank you @pkamra !

**Comment by scribwell** [comment]
@DalM2023 This tutorial is ready for editorial review.

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
+---
```
Comment (DalM2023) [inline tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md:1]: Hello! 

This blog has good structure, a friendly tone and what a fun project!! 

Please consider a few thing to keep in mind: 

- Please rake through the piece and uniform the spaces between sections, paragraphs and between paragraphs and instructions. Sometimes I saw one space, sometimes there's an extra space and sometimes there are no spaces. I tried to flag them but please do a double check so the spacing is uniform throughout the piece. 

- I made changes on the titles to AP Style, which is what Amazon uses. 

-  The capitalization of techniques and tools isn't uniform. Please be mindful of these. 
pdf = PDF
"in-context learning" has variations like "In-context learning". 
"vector database" is capitalized in some places and not in others.
I flagged as many as I could but please read through it to make sure they're uniform. 

- RE: LLM use. At the beginning of the piece you introduce LLM as a concept, title and acronym. So, there's no need to keep writing it as "Large Language Models (LLMs)" throughout the piece. You can simply use LLM. Also the M is model, so it's redundant to write "LLM model" and in plural it's LLMs not LLM's. 

Please see my notes below for other editorial suggestions. 

Overall, I enjoyed reading this blog. Well done! 👍

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+1. The very first step to provide In-Context learning is to ingest the pdf document and convert it into text chunks, generate vector representations of these text chunks called "embeddings" and finally store these embeddings in a vector database.
+
+2. Vector databases enable us to perform a "similarity search" against the text embeddings that are stored in it.
+
+3. [Amazon SageMaker JumpStart](https://docs.aws.amazon.com/sagemaker/latest/dg/studio-jumpstart.html) provides one click executable solution templates for setting up the infrastructure for pretrained, open-source models. We will be using Amazon SageMaker JumpStart to deploy the Embedding Model and the Large Language Model(LLM).<br/>
+[Amazon OpenSearch](https://aws.amazon.com/opensearch-service/) is a search and analytics engine that can search for nearest neighbors of points in a vector space, making it suitable as a Vector Database.
```
Comment (DalM2023) [inline tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md:43]: Does vector database need capitalization here?

**Review by DalM2023 — COMMENTED** [review]
1st round edit done. Please review.

**Comment by scribwell** [comment]
@pkamra We've finished the first pass of our editorial review. Please either accept our editor's changes or provide a reason for rejecting them and make any additional necessary edits by October 2. Thank you!

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+| ⏱ Time to complete    | 60 minutes                                                      |
+| 💰 Cost to complete    | Free tier eligible                                               |
+| 🧩 Prerequisites       | 1. You should have an active AWS account. If you don't have one, you can sign up on the [AWS website](https://aws.amazon.com/)<br/>2. Make sure you have the AWS Command Line Interface (CLI) installed on your local machine, and it should be properly configured with the necessary credentials and default region. You can configure it using the aws configure command.<br>3. Download and install Docker Engine. Follow the installation instructions for your specific operating system. |
+|  💻 Code Sample         | Code sample used in tutorial on [GitHub](https://github.com/build-on-aws/fullstack-llm-langchain-chatbot-on-aws)  |
+|
+
+| ToC |
+|-----|
+
+## What are we going to build
```
[the remark this answers is not in the document]
Comment (pkamra) [reply tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md:30]: Done

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+1. The very first step to provide In-Context learning is to ingest the pdf document and convert it into text chunks, generate vector representations of these text chunks called "embeddings" and finally store these embeddings in a vector database.
+
+2. Vector databases enable us to perform a "similarity search" against the text embeddings that are stored in it.
+
+3. [Amazon SageMaker JumpStart](https://docs.aws.amazon.com/sagemaker/latest/dg/studio-jumpstart.html) provides one click executable solution templates for setting up the infrastructure for pretrained, open-source models. We will be using Amazon SageMaker JumpStart to deploy the Embedding Model and the Large Language Model(LLM).<br/>
+[Amazon OpenSearch](https://aws.amazon.com/opensearch-service/) is a search and analytics engine that can search for nearest neighbors of points in a vector space, making it suitable as a Vector Database.
+
```
[the remark this answers is not in the document]
Comment (pkamra) [reply tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md:44]: Fixed

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+| 🧩 Prerequisites       | 1. You should have an active AWS account. If you don't have one, you can sign up on the [AWS website](https://aws.amazon.com/)<br/>2. Make sure you have the AWS Command Line Interface (CLI) installed on your local machine, and it should be properly configured with the necessary credentials and default region. You can configure it using the aws configure command.<br>3. Download and install Docker Engine. Follow the installation instructions for your specific operating system. |
+|  💻 Code Sample         | Code sample used in tutorial on [GitHub](https://github.com/build-on-aws/fullstack-llm-langchain-chatbot-on-aws)  |
+|
+
+| ToC |
+|-----|
+
+## What are we going to build
+In this example, we want to mimic the problem that is faced by many enterprises. Most of the data today is not structured but is rather unstructured in the form of audio and video transcripts, pdf and word documents, manuals, scanned notes, social media dumps, etc. We will use flan-t5-xxl model as the Large Language Model (LLM). This model can produce summaries, Q&A responses from unstructured texts. Here is the architecture showing the various building blocks.
```
[the remark this answers is not in the document]
Comment (pkamra) [reply tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md]: Done

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (pkamra) [reply tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md:44]: Fixed remaining white spaces between lines as well

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+  - sagemaker-jumpstart
+  - langchain
+  - tutorials
+showInHomeFeed: true
+authorGithubAlias: pkamra
+authorName: Piyali Kamra
+date: 2023-09-15
+---
+
+In today’s data-driven landscape, extracting meaningful insights from unstructured documents remains a challenge, inhibiting decision-making and innovation. Imagine being able to converse with your unstructured data, extracting valuable information with ease. In this tutorial, we will learn about embeddings, explore using Amazon OpenSearch as a vector database and seamlessly integrate [Langchain](https://python.langchain.com/docs/get_started/introduction/) framework with Large Language Models (LLMs) to build a website with an embedded NLP chatbot. We will go over the basics of LLM's to extract meaningful insights from an unstructured document with the help of an Open Source [Large Language Model](https://huggingface.co/google/flan-t5-xxl). By the end of this tutorial, you’ll have a comprehensive understanding of how to derive meaningful insights from unstructured documents, and use the skills to explore and innovate with similar full stack AI-driven solutions.
```
[the remark this answers is not in the document]
Comment (pkamra) [reply tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md:16]: Done

File: tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md (hunk 5 above)
Comment (pkamra) [reply tutorials/fullstack-llm-langchain-chatbot-on-aws/index.md:43]: fixed

**Comment by pkamra** [comment]
@scribwell @cherbk @DalM2023. Accepted and updated as per review comments.

**Review by DalM2023 — COMMENTED** [review]
@damonk10 Ready for review.

**Pull request merged**